Adding a new feature flag and setting to control SPE

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #3892

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

This PR introduces a new feature flag (`_wcstripe_feature_spe`) to control the Single Payment Element availability. This flag controls the feature as a whole, and does not behave like an option.

When the flag is enabled, and the legacy checkout experience is disabled, you should be able to toggle the "Single payment element" feature in settings. When the flag is disabled, the option is not displayed.

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- Checkout and build this branch on your test environment (`add/spe-feature-flag`)
- Connect your Stripe account
- Enable the new feature flag (`_wcstripe_feature_spe`). You can do it by either hardcoding the return value of `is_spe_available` to `true` or by running `npm run wp option update _wcstripe_feature_spe 'yes'`
- As a merchant, disable the legacy checkout experience in settings (`wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`)
- Confirm that you can see the new "Single payment element" setting:
<img width="757" alt="Screenshot 2025-02-27 at 10 40 54" src="https://github.com/user-attachments/assets/24315457-480e-401a-a35c-55e3a0f064a0" />

- Confirm that you can toggle it and that the value you set persists upon page refresh
- Disable the `_wcstripe_feature_spe` feature flag
- Confirm that the option is gone even when the legacy checkout experience is disabled

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Adding a new feature flag to control SPE

wjrosa committed:
Adding new setting to wp-admin

wjrosa committed:
Displaying SPE setting only when UPE is enabled

wjrosa committed:
Moving setting location

wjrosa committed:
Fix UPE dependency in settings

wjrosa committed:
Considering feature flag value when displaying the setting

wjrosa committed:
Fix tests

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Additional unit tests

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Merge branch 'develop' into add/spe-feature-flag

wjrosa committed:
Merge branch 'develop' into add/spe-feature-flag

Mayisha reviewed (commented):
@wjrosa thanks for adding this. The changes are mostly working except the saving part.

✅ SPE checkbox is displayed when the feature flag is enabled.
✅ SPE checkbox is hidden when the feature flag is disabled.
✅ SPE checkbox is disabled when legacy is enabled.
⚠️ Enabling the SPE checkbox does not persist on page refresh.

wjrosa committed:
Merge branch 'develop' into add/spe-feature-flag

wjrosa committed:
Fix setting persistence

wjrosa commented:
> @wjrosa thanks for adding this. The changes are mostly working except the saving part.
> 
> ✅ SPE checkbox is displayed when the feature flag is enabled. ✅ SPE checkbox is hidden when the feature flag is disabled. ✅ SPE checkbox is disabled when legacy is enabled. ⚠️ Enabling the SPE checkbox does not persist on page refresh.

Thanks for your review! I fixed the persistence issue in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3988/commits/f9596c9383e04fe422f7566bd239bb1b3cbcda17 . Also added tests for it

Mayisha reviewed (approved):
The settings persistence issue is fixed and working now 👍
